[TensorExpr] Cleanup implementation of alloc/free insertion.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35176 [TensorExpr] Cleanup implementation of alloc/free insertion.**
* #35175 [TensorExpr] Factor out LoopNest::insertAllocFree.
* #35174 [TensorExpr] Methods name cleanup in LoopNest class.
* #35173 [TensorExpr] Replace `ExprHandle` with `const Expr*` in `Substitute`.



Differential Revision: [D20585574](https://our.internmc.facebook.com/intern/diff/D20585574)

Commit:
[TensorExpr] Cleanup implementation of alloc/free insertion.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Cleanup implementation of alloc/free insertion."




Differential Revision: [D20585574](https://our.internmc.facebook.com/intern/diff/D20585574)

[ghstack-poisoned]

zheng-xq commented on torch/csrc/jit/tensorexpr/stmt.h:
```diff
@@ -112,6 +112,14 @@ class Block : public StmtNode<Block> {
   }
 
+  void prepend_stmt(Stmt* s) {
+    if (s->get_parent()) {
```
Why does this have to be a root? Also the error message here might be difficult to interpret.

ZolotukhinM replied:
This is how we're protecting against using statements in several other statements. When one inserts a statement into a block, the inserted statement must be without parent - it either needs to be created from scratch, or created by `Stmt::clone`. If you take a statement from existing block and try to insert it into another block, it will correctly fail at this check.

Also, please note that this PR does not introduce this mechanism, but rather follows already existing one (see `append_stmt`).

Commit:
Update on "[TensorExpr] Cleanup implementation of alloc/free insertion."




Differential Revision: [D20585574](https://our.internmc.facebook.com/intern/diff/D20585574)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Cleanup implementation of alloc/free insertion."




Differential Revision: [D20585574](https://our.internmc.facebook.com/intern/diff/D20585574)

[ghstack-poisoned]

Commit:
rebase on "[TensorExpr] Cleanup implementation of alloc/free insertion."




Differential Revision: [D20585574](https://our.internmc.facebook.com/intern/diff/D20585574)

[ghstack-poisoned]

Commit:
rebase on "[TensorExpr] Cleanup implementation of alloc/free insertion."




Differential Revision: [D20585574](https://our.internmc.facebook.com/intern/diff/D20585574)

[ghstack-poisoned]
